src/test/csf/Peer.h

```diff
@@ -306,146 +502,356 @@ struct Peer
-        consensus.peerProposal(now(), peerPos);
+    // Share a message by broadcasting to all connected peers
+    template <class M>
+    void
+    share(M const& m)
+    {
+        issue(Share<M>{m});
+        send(BroadcastMesg<M>{m,router.nextSeq++, this->id}, this->id);
     }
 
+    // Unwrap the Position and share the raw proposal
     void
-    receive(TxSet const& txs)
+    share(Position const & p)
     {
-        // save and map complete?
-        auto it = txSets.insert(std::make_pair(txs.id(), txs));
-        if (it.second)
-            consensus.gotTxSet(now(), txs);
+        share(p.proposal());
     }
 
+    //--------------------------------------------------------------------------
+    // Validation members
+
+    /** Add a trusted validation and return true if it is worth forwarding */
+    bool
+    addTrustedValidation(Validation v)
+    {
+        v.setTrusted();
+        v.setSeen(now());
+        AddOutcome res = validations.add(v.key(), v);
+
+        if(res == AddOutcome::stale || res == AddOutcome::repeat)
+            return false;
+
+        // Acquire will try to get from network if not already local
+        Ledger const* lgr = acquireLedger(v.ledgerID());
+        if (lgr)
+            checkFullyValidated(*lgr);
+        return true;
+    }
+
+    /** Check if a new ledger can be deemed fully validated */
     void
-    receive(Tx const& tx)
+    checkFullyValidated(Ledger const& ledger)
     {
-        if (openTxs.find(tx.id()) == openTxs.end())
+        // Only consider ledgers newer than our last fully validated ledger
+        if (ledger.seq() <= fullyValidatedLedger.seq())
+            return;
+
+        auto count = validations.numTrustedForLedger(ledger.id());
+        auto numTrustedPeers = trustGraph.graph().outDegree(this);
+        quorum = static_cast<std::size_t>(std::ceil(numTrustedPeers * 0.8));
+        if (count >= quorum)
         {
-            openTxs.insert(tx);
-            // relay to peers???
-            relay(tx);
+            issue(FullyValidateLedger{ledger, fullyValidatedLedger});
+            fullyValidatedLedger = ledger;
         }
     }
 
+    //-------------------------------------------------------------------------
+    // Peer messaging members
+
+    // Basic Sequence number router
+    //   A message that will be flooded across the network is tagged with a
+    //   seqeuence number by the origin node in a BroadcastMesg. Receivers will
+    //   ignore a message as stale if they've already processed a newer sequence
+    //   number, or will process and potentially relay the message along.
+    //
+    //  The various bool handle(MessageType) members do the actual processing
+    //  and should return true if the message should continue to be sent to peers.
+    //
+    //  WARN: This assumes messages are received and processed in the order they
+    //        are sent, so that a peer receives a message with seq 1 from node 0
+    //        before seq 2 from node 0, etc.
```

Comment:
Yes, I'm not sure yet how complex we want to model that layer. For now we have fixed delays, so it isn't an issue.  With changing delays, it would drop & fail to relay any that arrived out of order.